lib/Doctrine/Common/Inflector/Inflector.php

```diff
@@ -148,7 +176,76 @@ public static function reset() : void
      */
     public static function rules(string $type, iterable $rules, bool $reset = false) : void
     {
-        throw new BadMethodCallException('Adding custom rules is no longer supported in Doctrine Inflector 2.0.');
+        @trigger_error(sprintf('The "%s" method is deprecated and will be dropped in Doctrine Inflector 3.0. Please update to the new Inflector API.', __METHOD__), E_USER_DEPRECATED);
+
+        if (self::$factory === null) {
+            self::$factory = self::createFactory();
+        }
+
+        self::$instance = null;
+
+        switch ($type) {
+            case 'singular':
+                self::$factory->withSingularRules(self::buildRuleset($rules), $reset);
+                break;
+            case 'plural':
+                self::$factory->withPluralRules(self::buildRuleset($rules), $reset);
+                break;
+            default:
+                throw new InvalidArgumentException(sprintf('Cannot define custom inflection rules for type "%s".', $type));
+        }
+    }
+
+    private static function buildRuleset(iterable $rules) : Ruleset
+    {
+        $regular = [];
+        $irregular = [];
+        $uninflected = [];
+
+        foreach ($rules as $rule => $pattern) {
+            if ( ! is_array($pattern)) {
+                continue;
+            }
+
+            switch ($rule) {
+                case 'uninflected':
+                    $uninflected = $pattern;
+                    break;
+                case 'irregular':
+                    $irregular = $pattern;
+                    break;
+                case 'rules':
+                    $regular = $pattern;
+                    break;
+            }
+
+            unset($rules[$rule]);
+        }
+
+        $regular = (array) $rules + $regular;
```

Comment:
Casting a generator to an array will not work as expected I'm afraid: https://3v4l.org/8slDg

Reply:
Oh. So basically check if `$rules` is an iterator and call `iterator_to_array` on it, otherwise leave as is?

Reply:
Either that or use a `foreach` to merge `$rules` with `$regular`.

Reply:
Is this too much: https://3v4l.org/c3nTY ? You wouldn't need a foreach or a check for array or iterator though. My example is not preserving the keys, so you have to remove the `false` from `iterator_to_array`.

Reply:
The goal here is to add `$rules` to regular. I think a single foreach is the most concise way:

```
foreach ($rules as $rule => $pattern) {
    $regular[$rule] = $pattern;
}